### PR TITLE
fix(greptile-helper): suppress re-trigger when Greptile responded via in-place comment update

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Problem

When Greptile's formal PR review (`commit_id`) diverges from the current HEAD SHA, `_needs_re_review()` returns "re-review needed" indefinitely — even after Greptile has responded to a re-review trigger. This caused 3 triggers on #1246 in 2.5 hours (2026-07-09).

### Root cause

Greptile sometimes responds to `@greptileai review` triggers by updating its existing issue comment in-place rather than posting a new formal PR review with an updated `commit_id`. When this happens:

1. `_needs_re_review()` PRIMARY path: `reviewed_sha ≠ head_sha` → "re-review needed" (stays true forever)
2. Greptile's in-place comment update advances `reviewed_at` past the in-cycle trigger timestamp
3. `_no_new_commit_since_our_last_trigger` finds no in-cycle triggers (they're now "spent") → fails open
4. After the ack grace window, PM posts another trigger — every cycle, indefinitely

## Fix (v2 — head-SHA markers)

The first iteration compared Greptile's comment `updated_at` against the head commit's `committer.date`. Greptile's P1 correctly flagged that commit dates lag push time (written-locally-then-pushed commits look old — same failure class as gptme#2987), which masks genuinely-new heads. That version also broke two existing regression tests (`test_spent_trigger_with_backdated_commit_does_not_suppress`, `test_sha_mismatch_forces_re_review_even_when_date_says_no`); CI missed it due to path-filtered test jobs.

The fix now uses **server-side data only**:

- The `trigger` command embeds the PR head SHA in the trigger comment: `@greptileai review` + `<!-- greptile-helper head-sha: <sha> -->`
- In the SHA-mismatch branch, re-trigger is suppressed **only** when both hold:
  1. our latest trigger's marker SHA == current head SHA (the trigger was for exactly this head)
  2. Greptile's issue comment was updated *after* that trigger (a response landed)
- Legacy marker-less triggers keep pre-existing behavior ("re-review needed"); the next marker-carrying trigger self-heals the PR — at most one extra trigger per already-affected PR after deploy.

SHA equality + comment timestamps are both unforgeable server-side facts; no `committer.date` heuristics. (GraphQL `Commit.pushedDate` was considered but GitHub removed that field in 2023.)

## Testing

- All 27 tests pass, including the two regression tests the v1 approach broke
- 4 new tests: in-place-response suppression (#1246 case), backdated-new-head (the P1 case), unanswered-trigger non-suppression, marker embedding in posted triggers

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
Greptile's 4/5 concern is valid: `committer.date` is when a commit was made locally, not when it was pushed to GitHub. A commit made at T1 but pushed at T3, with Greptile responding at T2 (T1 < T2 < T3), would incorrectly suppress a needed re-trigger since T2 > T1.

**To fix and reach 5/5**, replace the REST `committer.date` lookup with GraphQL `pushedDate`:

```bash
# Current (line ~238-239 in _needs_re_review):
_sha_head_commit_date=$(gh api --paginate "repos/$REPO/pulls/$PR_NUMBER/commits" 2>/dev/null \
    | jq -rs '[.[][] | .commit.committer.date] | sort | last // ""' 2>/dev/null)

# Fixed: use GraphQL pushedDate (actual push time to GitHub)
_sha_head_commit_date=$(gh api graphql \
    -f owner="${REPO%%/*}" -f repo="${REPO##*/}" -F num="$PR_NUMBER" \
    -f query='query($owner:String!,$repo:String!,$num:Int!){repository(owner:$owner,name:$repo){pullRequest(number:$num){commits(last:1){nodes{commit{pushedDate}}}}}}' \
    2>/dev/null | jq -r '.data.repository.pullRequest.commits.nodes[0].commit.pushedDate // ""' 2>/dev/null)
```

The test's `FAKE_GH` also needs a `graphql` endpoint route added to the Python fake-gh stub, returning `pushedDate` from a `fixture.get("pushed_date", ...)` key.

In Bob's automated flow, this edge case is extremely rare (commits are almost always pushed immediately after creation). So merging at 4/5 is reasonable if the risk tolerance is acceptable. The edge case only bites when: a commit is pre-staged locally → Greptile reviews on a trigger for a prior commit → new commit is pushed. The formal fix above would close it completely.

File: scripts/github/greptile-helper.sh [hunk 1]
```diff
@@ -219,9 +219,29 @@ _needs_re_review() {
...
+            # advanced review_cutoff past the in-cycle trigger, causing
+            # _no_new_commit_since_our_last_trigger to fail open on every PM cycle.
+            local _sha_info _sha_reviewed_at _sha_head_commit_date
+            _sha_info=$(_greptile_review_info) || _sha_info=""
+            _sha_reviewed_at=$(echo "$_sha_info" | _json_field "reviewed_at") || _sha_reviewed_at=""
+            if [ -n "$_sha_reviewed_at" ] && [ "$_sha_reviewed_at" != "null" ]; then
+                _sha_head_commit_date=$(gh api --paginate "repos/$REPO/pulls/$PR_NUMBER/commits" 2>/dev/null \
+                    | jq -rs '[.[][] | .commit.committer.date] | sort | last // ""' 2>/dev/null) || _sha_head_commit_date=""
+                if [ -n "$_sha_head_commit_date" ] && _timestamp_gt "$_sha_reviewed_at" "$_sha_head_commit_date" 2>/dev/null; then
+                    return 1  # Issue comment updated after HEAD commit — Greptile responded, no re-trigger.
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply scripts/github/greptile-helper.sh]: Fixed in dabb4110 — and confirmed the finding was real beyond the theoretical case: the date heuristic broke two existing regression tests on this branch (`test_spent_trigger_with_backdated_commit_does_not_suppress` — the gptme#2987 case — and `test_sha_mismatch_forces_re_review_even_when_date_says_no` — the bob#890 case). CI missed it because the test-packages job was path-filter skipped.

The committer.date comparison is now gone entirely. Instead, trigger comments embed the head SHA they were posted for (`<!-- greptile-helper head-sha: ... -->`), and the SHA-mismatch branch suppresses only when our latest trigger was for the *current* head AND Greptile's issue comment was updated after that trigger — SHA equality + server-side comment timestamps, no commit dates involved. Legacy marker-less triggers keep pre-#1247 behavior and self-heal on the next marker-carrying trigger. 4 new regression tests cover the suppression, the backdated-head case from this finding, unanswered triggers, and marker embedding.

**Comment by ErikBjare** [comment]
Re the `pushedDate` suggestion above: that field can't be used — GitHub deprecated and removed `Commit.pushedDate` from the GraphQL API in 2023 (it returns null). dabb4110 takes a different route: trigger comments now embed the head SHA they were posted for, and the suppression check uses SHA equality + comment timestamps (all server-side facts) instead of any commit-date comparison. See the updated PR description and the thread reply on the P1 finding.

**Comment by ErikBjare** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review
